### PR TITLE
[router]: add new headless Tabs component

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

 Create new un-styled navigator components for Expo Router.

# How

Create new package export `expo-router/headless` 

## Components

```
<Tabs>
      <TabSlot /> // Renders the content of the tab
      <TabList> // Container view for the tab bar.
        <TabTrigger href="/ > // Tab pressable
          <Text>Home</Text>
        </TabTrigger>
        <TabTrigger href="/article">
            <Text>article</Text>
        </TabTrigger>
      </TabList>
    </Tabs>
```

`TabTrigger` supports the `asChild` props (similar to `<Link />` to allow for custom styled triggers.

<hr />

### `<Tabs>`

The container tabs. Internally this calls `useNavigationBuilder` and converts the children to `<Screen />`

```
type TabsProps = ViewProps & {
  options?: UseTabsOptions
};

// UseTabsOptions are the options passed to useNavigationBuilder, except initialRouteName and children
```

### `<TabSlot>`

```
export type TabSlotProps = ViewProps & {
  asChild?: boolean;
  options?: UseTabSlotOptions;
};
```

A `<Slot />` component that defaults to a `<View />`. A custom slot can be made with the `useTabSlot()` hook.

### `<TabList>`

A `<Slot />` component that defaults to a `<View />`. Only used as a wrapper for `<TabTriggers />`

### `<TabTrigger>`

A `<Slot />` component that defaults to a `<Link />`.

## Hooks

`useTabsWithChildren({ children })`

Can be used to create custom `<Tabs />` components. Will parse the `children` for triggers

`useTabsWithTriggers({ triggers })`

Can be used to create custom `<Tabs />` components. Triggers need to be explicitly passed. Useful when you need to customise how the `children` are rendered

`useTabsSlot(options?: UseTabsSlotOptions)`

```
export type UseTabsSlotOptions = {
  detachInactiveScreens?: boolean;
  renderFn?: ((descriptor: TabsDescriptor, { isFocused, loaded }) => React.JSX.Element) | undefined
};
```

A hook to render the current slot. Can pass a custom render function to customise the `<Screen />` (this is a `react-native-screens` screen)

# Test Plan

This is just the initial implementation with a unit test. Full tests coming in a later PR

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-router/src/useScreens.tsx [hunk 1]
```diff
@@ -266,26 +292,7 @@ function routeToScreen(route: RouteNode, { options, ...props }: Partial<ScreenPr
...
-        // Prevent generated screens from showing up in the tab bar.
-        if (route.generated) {
-          output.tabBarButton = () => null;
-          // TODO: React Navigation doesn't provide a way to prevent rendering the drawer item.
-          output.drawerItemStyle = { height: 0, display: 'none' };
-        }
-
-        return output;
-      }}
+      options={screenOptionsFactory(route, options)}
```
Comment (marklawlor) [inline packages/expo-router/src/useScreens.tsx:295]: Moving this into a new function to be reused in the headless components

File: packages/expo-router/src/headless/Tabs.slot.tsx [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+};
+
+export type TabsSlotRenderOptions = {
+  index: number;
+  isFocused: boolean;
+  loaded: boolean;
+};
+
+export function useTabSlot({
+  detachInactiveScreens = Platform.OS === 'web' ||
```
Comment (marklawlor) [inline packages/expo-router/src/headless/Tabs.slot.tsx:19]: This matches the logic in React Navigation. I'm not sure why the other out of tree platfoms are not allowed.

File: packages/expo-router/src/ui/TabTrigger.tsx [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        onPress={handleOnPress}
+        onLongPress={handleOnLongPress}
+        isFocused={isFocused(name)}>
+        {props.children}
+      </TabTriggerSlot>
+    );
+  } else {
+    return (
+      <Pressable
+        style={styles.tabTrigger}
```
Comment (EvanBacon) [inline packages/expo-router/src/ui/TabTrigger.tsx:84]: add `href`

File: packages/expo-router/src/ui/TabRouter.tsx [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        }
+      }
+
+      if (shouldReset) {
+        options.routeParamList[route.name] = {
+          ...options.routeParamList[route.name],
+          ...trigger.action.payload.params,
+        };
+        const state2 = rnTabRouter.getStateForAction(state, trigger.action, options);
+        return state2;
```
Comment (EvanBacon) [inline packages/expo-router/src/ui/TabRouter.tsx:89]:
```suggestion
        return rnTabRouter.getStateForAction(state, trigger.action, options);
```

File: packages/expo-router/src/layouts/withLayoutContext.tsx [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import { EventMapBase, NavigationState } from '@react-navigation/native';
 import React from 'react';
 
+import { View } from 'react-native';
+
```
Comment (EvanBacon) [inline packages/expo-router/src/layouts/withLayoutContext.tsx:5]:
```suggestion
```

File: apps/router-e2e/babel.config.js [hunk 6]
```diff
@@ -1,6 +1,10 @@
 module.exports = function (api) {
   api.cache(true);
+
+  const plugins = [];
+
```
Comment (EvanBacon) [inline apps/router-e2e/babel.config.js:5]:
```suggestion
```

File: apps/router-e2e/__e2e__/headless/app/+html.tsx [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+        {/* Test that public env vars are exposed */}
+        <meta name="expo-e2e-public-env-var" content={process.env.EXPO_PUBLIC_TEST_VALUE} />
+        {/* Test that server-only env vars are exposed as this file is a server file. */}
+        <meta name="expo-e2e-private-env-var" content={process.env.EXPO_NOT_PUBLIC_TEST_VALUE} />
+        <ScrollViewStyleReset />
+      </head>
+      <body>{children}</body>
+    </html>
+  );
+}
```
Comment (EvanBacon) [inline apps/router-e2e/__e2e__/headless/app/+html.tsx:28]: Delete this file if it's not critical

File: apps/router-e2e/package.json [hunk 8]
```diff
@@ -4,11 +4,14 @@
...
+    "start:01-rsc": "E2E_RSC_ENABLED=1 E2E_ROUTER_JS_ENGINE=hermes EXPO_USE_METRO_REQUIRE=1 E2E_CANARY_ENABLED=1 EXPO_USE_STATIC=single E2E_ROUTER_SRC=01-rsc EXPO_PUBLIC_USE_RSC=1 expo start",
+    "start:rsc": "E2E_RSC_ENABLED=1 E2E_ROUTER_JS_ENGINE=hermes EXPO_USE_METRO_REQUIRE=1 E2E_CANARY_ENABLED=1 EXPO_USE_STATIC=single E2E_ROUTER_SRC=rsc EXPO_PUBLIC_USE_RSC=1 expo start",
+    "export:rsc": "E2E_RSC_ENABLED=1 E2E_ROUTER_JS_ENGINE=hermes EXPO_USE_METRO_REQUIRE=1 E2E_CANARY_ENABLED=1 EXPO_USE_STATIC=single E2E_ROUTER_SRC=rsc EXPO_PUBLIC_USE_RSC=1 expo export",
+    "export:01-rsc": "E2E_RSC_ENABLED=1 E2E_ROUTER_JS_ENGINE=hermes EXPO_USE_METRO_REQUIRE=1 E2E_CANARY_ENABLED=1 EXPO_USE_STATIC=single E2E_ROUTER_SRC=01-rsc EXPO_PUBLIC_USE_RSC=1 expo export",
     "start:react-native-canary": "E2E_RSC_ENABLED=1 E2E_CANARY_ENABLED=1 E2E_ROUTER_SRC=react-native-canary expo start",
     "start:dom": "E2E_ROUTER_SRC=dom-components expo start -d",
     "dom:run:ios": "E2E_ROUTER_SRC=dom-components expo run:ios",
+    "headless:expo": "E2E_ROUTER_SRC=headless expo",
```
Comment (EvanBacon) [inline apps/router-e2e/package.json:14]:
```suggestion
    "start:headless": "E2E_ROUTER_SRC=headless expo",
```

please match the format of all the other scripts

**Review by EvanBacon — COMMENTED** [review]
please add documentation for the new API

**Pull request merged**